soci/ztoc_serializer.go

```diff
@@ -0,0 +1,244 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package soci
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sort"
+	"time"
+
+	ztoc_flatbuffers "github.com/awslabs/soci-snapshotter/soci/fbs/ztoc"
+	flatbuffers "github.com/google/flatbuffers/go"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// ZtocMarshaler incapsulates the serialization logic for Ztoc.
+// Currently, it implements serialization/deserialization to/from flatbuffers.
+type ZtocMarshaler struct {
+}
+
+// Marshal serializes Ztoc to its flatbuffers schema and returns a reader along with the descriptor (digest and size only).
+// If not successful, it will return an error.
+func (zm ZtocMarshaler) Marshal(ztoc *Ztoc) (io.Reader, ocispec.Descriptor, error) {
+	flatbuf, err := zm.ztocToFlatbuffer(ztoc)
+	if err != nil {
+		return nil, ocispec.Descriptor{}, err
+	}
+
+	buf := bytes.NewReader(flatbuf)
+	dgst := digest.FromBytes(flatbuf)
+	size := len(flatbuf)
+	return buf, ocispec.Descriptor{
+		Digest: dgst,
+		Size:   int64(size),
+	}, nil
+}
+
+// Unmarshal takes the reader with flatbuffers byte stream and deserializes it ztoc.
+// In case if there's any error situation during deserialization from flatbuffers, there will be an error returned.
+func (zm ZtocMarshaler) Unmarshal(serializedZtoc io.Reader) (*Ztoc, error) {
+	flatbuf, err := io.ReadAll(serializedZtoc)
+	if err != nil {
+		return nil, err
+	}
+
+	return zm.flatbufToZtoc(flatbuf)
+}
+
+func (zm ZtocMarshaler) flatbufToZtoc(flatbuffer []byte) (z *Ztoc, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			z = nil
+			err = fmt.Errorf("cannot unmarshal Ztoc")
```

Comment:
Can you add more context to the error message?

```
err = fmt.Errorf("cannot unmarshal ztoc: %v", r)
```

Reply:
Done.